homeassistant/helpers/device_registry.py

```diff
@@ -47,12 +47,15 @@ class DeletedDeviceEntry:
     identifiers: Set[Tuple[str, str]] = attr.ib()
     id: str = attr.ib()
 
-    def to_device_entry(self):
+    def to_device_entry(self, config_entry_id, connections, identifiers):
         """Create DeviceEntry from DeletedDeviceEntry."""
         return DeviceEntry(
-            config_entries=self.config_entries,
-            connections=self.connections,
-            identifiers=self.identifiers,
+            config_entries={config_entry_id},
+            connections=self.connections & connections,
+            identifiers=self.identifiers & identifiers,
+            inactive_config_entries=self.config_entries - {config_entry_id},
```

Comment:
Why not just throw it away? If a config entry comes in with a matching identifier/connection, it's added anyway?

Reply:
I was considering that, but I think the behavior will then appear very random to the users; sometimes `device_id` is stable and sometimes not.

Consider a scenario with devices identified by MAC, and where the user wants to test an alternative integration and then revert back.
An example is ESP devices which can be flashed with ESPHome or Tasmota, and where the Home Assistant integration may be ESPHome, MQTT or Tasmota.

If we keep the `inactive_xxx`, `device_id` is always guaranteed to be stable. If we throw them away, the behavior will appear a bit random to the user as it will depend on in which order things are removed and restored.

Reply:
Honestly, that's not a use case that we should care about supporting. We only have this functionality so that if you delete and add the same integration, your device IDs remain stable.

Reply:
OK, simplified accordingly.